### PR TITLE
Prepare HTTP middleware for Director

### PR DESCRIPTION
**Pull request by pkosiec**
**Description**

Changes proposed in this pull request:

- Prepare HTTP middleware for Director

**Related issue(s)**
#289

### CODE REVIEW

File: components/director/internal/authenticator/middleware_test.go [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+	}
+
+	return signedToken, nil
+}
+
+func createMiddleware(t *testing.T, allowJWTSigningNone bool) func(next http.Handler) http.Handler {
+	auth := authenticator.New(PublicJWKSURL, allowJWTSigningNone)
+	publicJWKS, err := authenticator.FetchJWK(PublicJWKSURL)
+	require.NoError(t, err)
+	auth.Jwks = publicJWKS
```
Comment (aszecowka) [inline components/director/internal/authenticator/middleware_test.go]: why you don't call `SynchronizeJWKS` ?

File: components/director/internal/authenticator/middleware_test.go [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+
+	signedToken, err := token.SignedString(jwt.UnsafeAllowNoneSignatureType)
+	if err != nil {
+		return "", errors.Wrap(err, "while signing token")
+	}
+
+	return signedToken, nil
+}
+
+func createTokenWithSigningMethod(t *testing.T, tnt string, scopes []string, key jwk.Key) (string, error) {
```
Comment (aszecowka) [inline components/director/internal/authenticator/middleware_test.go]: don't return an error, you have access to `testing.T` and you can validate errors internally

File: components/director/internal/authenticator/middleware_test.go [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+
+}
+
+type jwtTokenClaims struct {
+	Scopes []string `json:"scopes"`
+	Tenant string   `json:"tenant"`
+	jwt.StandardClaims
+}
+
+func createToken(t string, scopes []string) (string, error) {
```
Comment (aszecowka) [inline components/director/internal/authenticator/middleware_test.go]: maybe `createNotSingedToken`?

Comment (aszecowka) [inline components/director/internal/authenticator/middleware_test.go]: 1. what is `t` - usually it is `testing.T`
2. in other places you have testing.T and internally validate errors, please make it consistent

File: components/director/internal/authenticator/middleware_test.go [hunk 4]
```diff
@@ -0,0 +1,277 @@
+package authenticator_test
```
Comment (aszecowka) [inline components/director/internal/authenticator/middleware_test.go]: code coverage is 92%, do you have all important test cases covered?

File: components/director/internal/authenticator/middleware.go [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+	}
+}
+
+func (a *Authenticator) getBearerToken(r *http.Request) (string, error) {
+	reqToken := r.Header.Get(AuthorizationHeaderKey)
+	if reqToken == "" {
+		return "", errors.New("Invalid bearer token")
+	}
+	splitToken := strings.Split(reqToken, "Bearer ")
+	reqToken = splitToken[1]
```
Comment (aszecowka) [inline components/director/internal/authenticator/middleware.go]: can we get here `Index Out of Range` err?

File: components/director/internal/authenticator/middleware.go [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+
+func (a *Authenticator) contextWithClaims(ctx context.Context, claims Claims) context.Context {
+	ctxWithTenant := tenant.SaveToContext(ctx, claims.Tenant)
+	ctxWithScopes := scope.SaveToContext(ctxWithTenant, claims.Scopes)
+	return ctxWithScopes
+}
+
+func (a *Authenticator) getKeyFunc() func(token *jwt.Token) (interface{}, error) {
+	return func(token *jwt.Token) (interface{}, error) {
+		unsupportedErr := fmt.Errorf("Unexpected signing method: %v", token.Method.Alg())
```
Comment (aszecowka) [inline components/director/internal/authenticator/middleware.go]: error string should not be capitalized, the same below

File: components/director/internal/authenticator/jwks_fetch.go [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+		return jwk.FetchHTTP(urlstring, options...)
+	case "file":
+		pathPart := strings.Split(urlstring, "file://")
+		if len(pathPart) < 2 {
+			return nil, errors.New("Incorrect file path")
+		}
+
+		f, err := os.Open(pathPart[1])
+		if err != nil {
+			return nil, errors.Wrap(err, `failed to open jwk file`)
```
[the remark this answers is not in the document]
Comment (pkosiec) [reply components/director/internal/authenticator/jwks_fetch.go]: This is copied from library

File: components/director/internal/authenticator/jwks_fetch.go [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+	u, err := url.Parse(urlstring)
+	if err != nil {
+		return nil, errors.Wrap(err, `failed to parse url`)
+	}
+
+	switch u.Scheme {
+	case "http", "https":
+		return jwk.FetchHTTP(urlstring, options...)
+	case "file":
+		pathPart := strings.Split(urlstring, "file://")
```
[the remark this answers is not in the document]
Comment (pkosiec) [reply components/director/internal/authenticator/jwks_fetch.go]: tenor-86289622

I'll use `TrimPrefix`

File: components/director/internal/authenticator/middleware.go (hunk 5 above)
Comment (pkosiec) [reply components/director/internal/authenticator/middleware.go]: Not anymore, I've used `TrimPrefix`

File: components/director/internal/authenticator/middleware_test.go (hunk 4 above)
Comment (pkosiec) [reply components/director/internal/authenticator/middleware_test.go]: yes

**Pull request merged**